api/src/server/auth.rs

```diff
@@ -0,0 +1,140 @@
+    }
+}
+
+// The frontend will redirect to this before being sent off to the
+// actual openid provdider
+// TODO: add query param to say which provider to use
+#[get("/oauth2/authorization/oidc")]
+pub async fn route_authorize(
+    oidc_client: web::Data<DiscoveredClient>,
+) -> impl Responder {
+    let auth_url = oidc_client.auth_url(&Options {
+        scope: Some("email".into()),
+        ..Default::default()
+    });
+
+    HttpResponse::Found()
+        .header(http::header::LOCATION, auth_url.to_string())
+        .finish()
+}
+
+#[derive(Deserialize, Debug)]
+pub struct LoginQuery {
+    code: String,
+}
+
+// Exchanges the access token from the inital login in the openid provider
+// for a normal token
+async fn request_token(
+    oidc_client: web::Data<DiscoveredClient>,
+    query: web::Query<LoginQuery>,
+) -> Result<Option<(Token, Userinfo)>, actix_web::Error> {
+    let mut token: Token = oidc_client
+        .request_token(&query.code)
+        .await
+        .map_err(ResponseError::from)?
+        .into();
+    if let Some(mut id_token) = token.id_token.as_mut() {
+        oidc_client
+            .decode_token(&mut id_token)
+            .map_err(ResponseError::from)?;
+        oidc_client
+            .validate_token(&id_token, None, None)
+            .map_err(ResponseError::from)?;
+    } else {
+        return Ok(None);
+    }
+    let userinfo = oidc_client
+        .request_userinfo(&token)
+        .await
+        .map_err(ResponseError::from)?;
+
+    Ok(Some((token, userinfo)))
+}
+
+fn host(path: &str) -> String {
+    "http://localhost:8000".to_string() + path
+}
+
+// Provider redirects back to this route after the login
+#[get("/login/oauth2/code/oidc")]
+pub async fn route_login(
+    oidc_client: web::Data<DiscoveredClient>,
+    query: web::Query<LoginQuery>,
+    sessions: web::Data<RwLock<Sessions>>,
+    pool: web::Data<Pool>,
+    identity: Identity,
+) -> Result<HttpResponse, actix_web::Error> {
+    let conn = &pool.get().map_err(ResponseError::from)? as &PgConnection;
+
+    match request_token(oidc_client, query).await {
+        Ok(Some((token, userinfo))) => {
+            let email: String = userinfo.email.clone().unwrap();
+            // TODO: handle setting username
+            let user: User = NewUser {
+                username: &email.chars().take(20).collect::<String>(),
+            }
+            .insert()
+            .returning(users::all_columns)
+            .get_result(conn)
+            .unwrap();
+            let id = user.id.to_string();
+
+            identity.remember(id.clone());
+            sessions
+                .write()
+                .unwrap()
+                .map
+                .insert(id, (user, token, userinfo));
+            // TODO: add redirect path to state param so we dont always redirect
+            // to the homepage
+            Ok(HttpResponse::Found()
+                .header(http::header::LOCATION, "http://localhost:3000/")
```

Comment:
This should redirect to just `/`. If you add the `/api/` to the start of the routes, then you can access them through the proxy on 3000, so that you can redirect to FE routes without needing the hostname.

Reply:
It might be a little weird since the provider redirects to `localhost:8000/api/callback` so idk if the proxy will pick that up, ill try though

Reply:
yea when you do `/` it just goes to `localhost:8000/`